File: spark-on-angel/core/src/main/scala/com/tencent/angel/spark/linalg/Vector.scala
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.tencent.angel.spark.linalg

import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap

trait Vector extends Serializable {
  def length: Long
  def apply(index: Long): Double

  def toDense: DenseVector

  def toSparse: SparseVector
}

class DenseVector(val values: Array[Double]) extends Vector {
  def length: Long = values.length

  def apply(index: Long): Double = values(index.toInt)

  def toDense: DenseVector = this

  def toSparse: SparseVector = {
    val nzIndices = new Array[Long](nnz)
    val nzValues = new Array[Double](nnz)
    var i = 0
    values.indices.foreach { j =>
      if (values(j) != 0.0) {
        nzIndices(i) = j.toLong
        nzValues(i) = values(j)
        i += 1
      }
    }
    new SparseVector(length, nzIndices, nzValues)
  }

  def toArray: Array[Double] = values

  def nnz: Int = {
    var nonZeroNum = 0
    values.foreach { x =>
      if (x != 0) nonZeroNum += 1
    }
    nonZeroNum
  }
}

class SparseVector(override val length: Long) extends Vector {
  var keyValues: Long2DoubleOpenHashMap = new Long2DoubleOpenHashMap()

  def this(len: Long, indices: Array[Long], values: Array[Double]) {
    this(len)
    keyValues = null
    keyValues = new Long2DoubleOpenHashMap(indices, values)
  }

  def this(len: Long, pairs: Array[(Long, Double)]) {
    this(len)
    val (keys, values) = pairs.unzip
    keyValues = null
    keyValues = new Long2DoubleOpenHashMap(keys, values)
  }

  def this(len: Long, other: Long2DoubleOpenHashMap) {
    this(len)
    keyValues = null
    keyValues = other.clone()
  }

  def this(len: Long, expectedSize: Int) {
    this(len)
    keyValues = null
    keyValues = new Long2DoubleOpenHashMap(expectedSize)
  }

  def reSize(size: Long): Unit = {
    val temp = new Long2DoubleOpenHashMap(size.toInt)
    temp.putAll(keyValues)
    keyValues = temp
  }

  def toDense: DenseVector = throw new Exception("Can not convert SparseVector to DenseVector")

  def toSparse: SparseVector = this

  def nnz: Long = {
    import scala.collection.JavaConverters._
    var num = 0L
    for (entry <- keyValues.asScala) {
      if (math.abs(entry._2 - keyValues.defaultReturnValue()) > 1e-11) {
        num += 1
      }
    }
    num
  }

  def apply(index: Long): Double = {
    keyValues.get(index)
  }

  def indices: Array[Long] = {
    keyValues.keySet().toLongArray
  }

  def values: Array[Double] = {
    keyValues.values().toDoubleArray
  }

  def pairs: (Long, Array[Long], Array[Double]) = {
    require(keyValues.defaultReturnValue() == 0.0, "SparseVector default value must be 0.0")
    val keys = new Array[Long](keyValues.size())
    val values = new Array[Double](keyValues.size())
    import scala.collection.JavaConverters._
    var i: Int = 0
    for (entry <- keyValues.asScala) {
      keys(i) = entry._1
      values(i) = entry._2
      i += 1
    }
    (length, keys, values)
  }

  def put(key: Long, value: Double): Unit = {
    keyValues.put(key, value)
  }
}

class OneHotVector(override val length: Long) extends Vector {
  var indices: Array[Long] = _

  def this(length: Long, ids: Array[Long]) {
    this(length)
    require(ids.length <= length)
    indices = ids.sorted
  }

  def this(length: Long, indices: Array[Int]) {
    this(length, indices.map(_.toLong))
  }

  def apply(index: Long): Double = {
    if (java.util.Arrays.binarySearch(indices, index) >= 0) 1.0 else 0.0
  }

  def toDense: DenseVector = throw new Exception("Can not convert OneHotVector to DenseVector")

  def toSparse: SparseVector = {
    val values = Array.fill(indices.length)(1.0)
    new SparseVector(length, indices, values)
  }
}

File: angel-ps/core/src/main/java/com/tencent/angel/ml/math2/vector/Vector.java
/*
 * Tencent is pleased to support the open source community by making Angel available.
 *
 * Copyright (C) 2017-2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in 
 * compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/Apache-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 *
 */


package com.tencent.angel.ml.math2.vector;

import com.tencent.angel.ml.math2.MathObject;
import com.tencent.angel.ml.math2.storage.IntKeyVectorStorage;
import com.tencent.angel.ml.math2.storage.LongKeyVectorStorage;
import com.tencent.angel.ml.math2.storage.VectorStorage;
import com.tencent.angel.ml.math2.ufuncs.executor.BinaryExecutor;
import com.tencent.angel.ml.math2.ufuncs.executor.DotExecutor;
import com.tencent.angel.ml.math2.ufuncs.executor.UnaryExecutor;
import com.tencent.angel.ml.math2.ufuncs.expression.*;
import com.tencent.angel.ml.matrix.RowType;
import org.apache.commons.lang.NotImplementedException;

import java.io.Serializable;

public abstract class Vector extends MathObject implements Serializable {
  protected int rowId;

  protected VectorStorage storage;

  public void setStorage(VectorStorage storage) {
    this.storage = storage;
  }

  public VectorStorage getStorage() {
    return storage;
  }

  public void clear() {
    storage.clear();
  }

  public int getRowId() {
    return rowId;
  }

  public void setRowId(int rowId) {
    this.rowId = rowId;
  }

  public boolean isDense() {
    return storage.isDense();
  }

  public boolean isSparse() {
    return storage.isSparse();
  }

  public boolean isSorted() {
    return storage.isSorted();
  }

  public double dot(Vector other) {
    return DotExecutor.apply(this, other);
  }

  public Vector iadd(Vector other) {
    return BinaryExecutor.apply(this, other, new Add(true));
  }

  public Vector add(Vector other) {
    return BinaryExecutor.apply(this, other, new Add(false));
  }

  public Vector isub(Vector other) {
    return BinaryExecutor.apply(this, other, new Sub(true));
  }

  public Vector sub(Vector other) {
    return BinaryExecutor.apply(this, other, new Sub(false));
  }

  public Vector imul(Vector other) {
    return BinaryExecutor.apply(this, other, new Mul(true));
  }

  public Vector mul(Vector other) {
    return BinaryExecutor.apply(this, other, new Mul(false));
  }

  public Vector idiv(Vector other) {
    return BinaryExecutor.apply(this, other, new Div(true));
  }

  public Vector div(Vector other) {
    return BinaryExecutor.apply(this, other, new Div(false));
  }

  public Vector iaxpy(Vector other, double alpha) {
    return BinaryExecutor.apply(this, other, new Axpy(true, alpha));
  }

  public Vector axpy(Vector other, double alpha) {
    return BinaryExecutor.apply(this, other, new Axpy(false, alpha));
  }

  public Vector iadd(double x) {
    return UnaryExecutor.apply(this, new SAdd(true, x));
  }

  public Vector add(double x) {
    return UnaryExecutor.apply(this, new SAdd(false, x));
  }

  public Vector isub(double x) {
    return UnaryExecutor.apply(this, new SSub(true, x));
  }

  public Vector sub(double x) {
    return UnaryExecutor.apply(this, new SSub(false, x));
  }

  public Vector imul(double x) {
    return UnaryExecutor.apply(this, new SMul(true, x));
  }

  public Vector mul(double x) {
    return UnaryExecutor.apply(this, new SMul(false, x));
  }

  public Vector idiv(double x) {
    return UnaryExecutor.apply(this, new SDiv(true, x));
  }

  public Vector div(double x) {
    return UnaryExecutor.apply(this, new SDiv(false, x));
  }

  public abstract Vector filter(double threshold);

  public abstract Vector ifilter(double threshold);

  public abstract Vector filterUp(double threshold);

  public RowType getType() {
    if (null != storage) {
      return storage.getType();
    } else {
      return this.getType();
    }
  }

  public long getSize() {
    if (storage instanceof IntKeyVectorStorage) {
      return (long) ((IntKeyVectorStorage) storage).size();
    } else if (storage instanceof LongKeyVectorStorage) {
      return ((LongKeyVectorStorage) storage).size();
    } else if (this instanceof IntKeyVector) {
      return (long) ((IntKeyVector) this).size();
    } else if (this instanceof LongKeyVector) {
      return ((LongKeyVector) this).size();
    } else {
      throw new NotImplementedException();
    }
  }

  public abstract Vector copy();

  public abstract long dim();
}
